File: BOJ/17142/17142.java
import java.io.*;
import java.util.*;

class Main{
    private static class Info {
        int x, y;
        private Info(int i, int j) {
            this.x = i;
            this.y = j;
        }
    }
    private static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    private static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

    private static StringTokenizer st;
    private static String[] token;
    private static int N, M, totalBlank, answer;
    private static final int MAX = 987654321;
    private static final int BLANK = 0;
    private static final int BLOCK = 1;
    private static final int VIRUS = 2;
    private static final int DONE = 3;
    private static final int LEN = 50;
    private static final int[][] dirs = {{1,0},{-1,0},{0,1},{0,-1}};
    private static int[][] A = new int[LEN + 1][LEN + 1];
    private static int[][] B = new int[LEN + 1][LEN + 1];

    private static ArrayList<Info> viruses = new ArrayList<>();
    private static ArrayList<Info> cur = new ArrayList<>();
    static int pI(String s) {
        return Integer.parseInt(s);
    }
    private static void init() throws IOException {
        token = br.readLine().split(" ");
        N = pI(token[0]);
        M = pI(token[1]);
        viruses.clear();
        cur.clear();
        totalBlank = 0;
        answer = MAX;
        for (int i = 0; i < N; i++) {
            token = br.readLine().split(" ");
            for (int j = 0; j < N; j++) {
                A[i][j] = pI(token[j]);
                if (A[i][j] == VIRUS) {
                    viruses.add(new Info(i, j));
                }
                if (A[i][j] == BLANK) totalBlank++;
            }
        }
    }
    private static int bfs() throws Exception {
        Queue<Info> q = new LinkedList<>();
        int blankCnt = 0, cnt = 0;
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < N; j++) {
                B[i][j] = A[i][j];
                if (B[i][j] == BLANK) blankCnt++;
            }
        }
        for (Info e : cur) {
            B[e.x][e.y] = DONE;
            q.add(new Info(e.x, e.y));
        }
        while (!q.isEmpty()) {
            Queue<Info> temp_q = new LinkedList<>();
            while (!q.isEmpty()) {
                Info cur = q.poll();
                for (int[] d : dirs) {
                    int nx = cur.x + d[0];
                    int ny = cur.y + d[1];
                    if (nx < 0 || ny < 0 || nx >= N || ny >= N) continue;
                    if (B[nx][ny] == DONE || B[nx][ny] == BLOCK) continue;
                    if (B[nx][ny] == BLANK) blankCnt--;
                    B[nx][ny] = DONE;
                    temp_q.add(new Info(nx, ny));
                }
            }
            q = temp_q;
            if (!q.isEmpty()) cnt++;
            if (blankCnt == 0) break;
        }
        return blankCnt == 0 ? cnt : MAX;
    }
    private static void go(int idx) throws Exception {
        if (cur.size() == M) {
            answer = Math.min(answer, bfs());
            return;
        }
        for (int i = idx; i < viruses.size(); i++) {
            cur.add(viruses.get(i));
            go(i + 1);
            cur.remove(cur.size() - 1);
        }
    }
    public static void main(String[] args) throws Exception {
        init();
        if (totalBlank == 0) {
            bw.write("0 \n");
        }
        else {
            go(0);
            if (answer == MAX) bw.write("-1 \n");
            else bw.write("" + answer);
        }
        bw.flush();
    }
}

File: BOJ/17142/17142.py
import sys
from collections import deque
input = sys.stdin.readline

MAX, BLANK, BLOCK, VIRUS, DONE = 987654321, 0, 1, 2, 3
N, M = map(int, input().split())
A = [list(map(int, input().split())) for _ in range(N)]
B = [[0 for _ in range(N)] for _ in range(N)]
dirs = [[1,0],[-1,0],[0,1],[0,-1]]
answer, totalBlank = MAX, 0

viruses = []

for i in range(N):
    for j in range(N):
        if A[i][j] == VIRUS:
            viruses.append([i, j])
        if A[i][j] == BLANK:
            totalBlank += 1

def bfs() -> int:
    q = deque()
    blankCnt, cnt = 0, 0
    for r in range(N):
        for c in range(N):
            B[r][c] = A[r][c]
            if B[r][c] == BLANK:
                blankCnt += 1
    for e in cur:
        q.append([viruses[e][0], viruses[e][1]])
        B[viruses[e][0]][viruses[e][1]] = DONE
    while q:
        temp_q = deque()
        for x, y in q:
            for d in dirs:
                nx, ny = x + d[0], y + d[1]
                if nx < 0 or ny < 0 or nx >= N or ny >= N:
                    continue
                if B[nx][ny] == DONE or A[nx][ny] == BLOCK:
                    continue
                if B[nx][ny] == BLANK:
                    blankCnt -= 1
                B[nx][ny] = DONE
                temp_q.append([nx, ny])
        q = temp_q
        if q:
            cnt += 1
        if not blankCnt:
            break

    return MAX if blankCnt > 0 else cnt

cur = []
def go(idx):
    if len(cur) == M:
        global answer
        answer = min(answer, bfs())
        return
    for i in range(idx, len(viruses)):
        cur.append(i)
        go(i + 1)
        cur.pop()

if totalBlank == 0:
    print(0)
else:
    go(0)
    print(answer if answer != MAX else -1)